collector/src/main/java/io/prometheus/jmx/JmxScraper.java

```diff
@@ -108,17 +96,19 @@ public void doScrape() throws Exception {
         }
         try {
             // Query MBean names, see #89 for reasons queryMBeans() is used instead of queryNames()
-            Set<ObjectInstance> mBeanNames = new HashSet();
+            Set<ObjectInstance> mBeanInstances = new HashSet<ObjectInstance>();
             for (ObjectName name : whitelistObjectNames) {
-                mBeanNames.addAll(beanConn.queryMBeans(name, null));
+                mBeanInstances.addAll(beanConn.queryMBeans(name, null));
             }
             for (ObjectName name : blacklistObjectNames) {
-                mBeanNames.removeAll(beanConn.queryMBeans(name, null));
+                Set<ObjectInstance> queriedMBeans = beanConn.queryMBeans(name, null);
+                jmxMBeanPropertyCache.removeMBeans(queriedMBeans);
```

Comment:
@brian-brazil I tried to cover the memleak here. Did you have something else in mind?

Reply:
The leak is when the mBeans change from scrape to scrape, which this doesn't cover.

Reply:
Gotcha; I think I can fix this by keeping a list of queryMBeans that get collected via whilelist loop above. I'd compare that list to what it was before and call removeMBean for anything still in the list from previous run.

Reply:
See my updated change; I clear the mBeans if they are no longer in whitelist; I still need to test this out though. Is a good way to test this case to remove things from whitelist in the config and verify that mBean gets removed? I'll write up some utests for this too if you consider the approach agreeable. 

Do you have an app that has mBeans chaning from scrape-to-space that I can test with?